cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -416,27 +420,38 @@ private void doPerformAction(FullHttpRequest request, HttpResponder responder, S
     ProgramType programType = getProgramType(type);
     ProgramId program = applicationId.program(programType, programId);
     Map<String, String> args = decodeArguments(request);
-    String latestVersion = getLatestAppVersion(new NamespaceId(namespaceId), appId);
     // we have already validated that the action is valid
     switch (action.toLowerCase()) {
       case "start":
-        if (!appVersion.equals(latestVersion)) {
-          throw new BadRequestException("start action is only allowed on the latest program version");
+        if (ApplicationId.DEFAULT_VERSION.equals(appVersion)) {
+          lifecycleService.run(new NamespaceId(namespaceId), appId, programType, programId, args, false);
+        } else {
+          lifecycleService.run(program, args, false);
         }
-        lifecycleService.run(program, args, false);
         break;
       case "debug":
-        if (!appVersion.equals(latestVersion)) {
-          throw new BadRequestException("debug action is only allowed on the latest program version");
-        }
         if (!isDebugAllowed(programType)) {
           throw new NotImplementedException(String.format("debug action is not implemented for program type %s",
                                                           programType));
         }
-        lifecycleService.run(program, args, true);
+        if (ApplicationId.DEFAULT_VERSION.equals(appVersion)) {
+          lifecycleService.run(new NamespaceId(namespaceId), appId, programType, programId, args, true);
+        } else {
+          lifecycleService.run(program, args, true);
+        }
         break;
       case "stop":
-        lifecycleService.stop(program);
+        if (ApplicationId.DEFAULT_VERSION.equals(appVersion)) {
+          lifecycleService.stop(new NamespaceId(namespaceId), appId, programType, programId);
+          // handle the edge case where both -SNAPSHOT and latest version are running and stop them both
```

Comment:
I don't think we need to do this anymore, as we've decided -snapshot will always mean latest, and will only be visible in the /apps/{app-id}/versions list if it is actually the latest.

Reply:
removed